src/states/elder/mod.rs

```diff
@@ -1394,20 +1346,74 @@ impl Base for Elder {
         Ok(Transition::Stay)
     }
 
-    fn unhandled_message(&mut self, _sender: Option<SocketAddr>, _msg: Message) {
-        unreachable!()
+    fn unhandled_message(&mut self, sender: Option<SocketAddr>, msg: Message, msg_bytes: Bytes) {
+        match msg.variant {
+            Variant::GenesisUpdate(_) | Variant::Relocate(_) | Variant::MessageSignature(_) => {
+                debug!("{} Unhandled message, bouncing: {:?}", self, msg);
+                self.send_bounce(sender, msg_bytes);
+            }
+            Variant::BootstrapResponse(_) | Variant::NodeApproval(_) | Variant::Ping => {
+                debug!("{} Unhandled message, discarding: {:?}", self, msg);
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn is_message_handled(&self, msg: &MessageWithBytes) -> bool {
+        self.msg_filter.contains_incoming(msg)
     }
 
-    fn filter_incoming_message(&mut self, message: &MessageWithBytes) -> bool {
-        self.msg_filter.filter_incoming(message).is_new()
+    fn set_message_handled(&mut self, msg: &MessageWithBytes) {
+        self.msg_filter.insert_incoming(msg)
     }
 
-    fn relay_message(&mut self, message: &MessageWithBytes) -> Result<()> {
+    fn relay_message(
+        &mut self,
+        _sender: Option<SocketAddr>,
+        message: &MessageWithBytes,
+    ) -> Result<()> {
         self.send_signed_message(message)
     }
 
-    fn should_handle_message(&self, _msg: &Message) -> bool {
-        true
+    fn should_handle_message(&self, msg: &Message) -> bool {
+        match &msg.variant {
+            Variant::NeighbourInfo(_)
+            | Variant::UserMessage(_)
+            | Variant::AckMessage { .. }
+            | Variant::BootstrapRequest(_)
+            | Variant::JoinRequest(_)
+            | Variant::MemberKnowledge(_)
+            | Variant::ParsecRequest(..)
+            | Variant::ParsecResponse(..)
+            | Variant::Bounce { .. } => true,
+
+            Variant::MessageSignature(accumulating_msg) => {
+                match &accumulating_msg.content.variant {
+                    Variant::NeighbourInfo(_)
+                    | Variant::UserMessage(_)
+                    | Variant::NodeApproval(_)
+                    | Variant::AckMessage { .. }
+                    | Variant::Relocate(_) => true,
+
+                    Variant::GenesisUpdate(_)
+                    | Variant::MessageSignature(_)
+                    | Variant::BootstrapRequest(_)
+                    | Variant::BootstrapResponse(_)
+                    | Variant::JoinRequest(_)
+                    | Variant::MemberKnowledge(_)
+                    | Variant::ParsecRequest(..)
+                    | Variant::ParsecResponse(..)
+                    | Variant::Ping
+                    | Variant::Bounce { .. } => false,
```

Comment:
We do handle `Bounce` (see line 1394). Here we just ignore `Bounce` that is inside `MessageSignature` which would be invalid anyway because `Bounce` doesn't use signature accumulation.